From akinobu.mita@gmail.com Sun Feb 17 01:20:20 2008
Return-Path: <akinobu.mita@gmail.com>
Received: from imap.linux.ibm.com ([unix socket]) by imap.linux.ibm.com
	(Cyrus v2.3.7-Invoca-RPM-2.3.7-7) with LMTPA; Sun, 17 Feb 2008 01:20:20
	-0500
X-Sieve: CMU Sieve 2.3
Received: by imap.linux.ibm.com (Postfix, from userid 101) id EE3481910146;
	Sun, 17 Feb 2008 01:20:20 -0500 (EST)
X-Spam-TestScore: SPF_NEUTRAL=1.379
X-Spam-TokenSummary: Bayes not run.
X-Spam-Checker-Version: SpamAssassin 3.1.7 (2006-10-05) on
	imap.linux.ibm.com
X-Spam-Level: *
X-Spam-Status: No, score=1.4 required=5.0 tests=SPF_NEUTRAL 
	autolearn=disabled version=3.1.7
X-Spam-Relay-Country: US ** US US US ** US US XX JP
Received: from smtp.linux.ibm.com (smtp.linux.ibm.com [9.26.4.197]) by
	imap.linux.ibm.com (Postfix) with ESMTP id 927B8191009F for
	<cmm@imap.linux.ibm.com>; Sun, 17 Feb 2008 01:20:16 -0500 (EST)
Received: from localhost (localhost.localdomain [127.0.0.1]) by
	smtp.linux.ibm.com (Postfix) with ESMTP id 4CADFC03F for
	<cmm@linux.ibm.com>; Sun, 17 Feb 2008 01:20:16 -0500 (EST)
X-Virus-Scanned: amavisd-new at linux.ibm.com
Received: from BLDVMB.POK.IBM.COM (bldvmb.pok.ibm.com [9.57.6.113]) by
	smtp.linux.ibm.com (Postfix) with ESMTP id F3F95C03E for
	<cmm@linux.ibm.com>; Sun, 17 Feb 2008 01:20:15 -0500 (EST)
Received:  by BLDVMB.POK.IBM.COM (IBM VM SMTP Level 440) via spool with
	SMTP id 2286 ; Sat, 16 Feb 2008 23:21:17 MST
Received: by bldvmb.vnet.ibm.com (xagent2 5.9.5) via xagsmtp3 with spool id
	2462 for cmm@linux.vnet.ibm.com; Sat, 16 Feb 2008 23:21:17 -0700 (MST)
Received: from d03relay05.boulder.ibm.com [9.17.195.107] by
	BLDVMB.POK.IBM.COM (IBM VM SMTP Level 440) via TCP with ESMTP ; Sat, 16 Feb
	2008 23:21:16 MST
Received: from d03av03.boulder.ibm.com (d03av03.boulder.ibm.com
	[9.17.195.169]) by d03relay05.boulder.ibm.com (8.13.8/8.13.8/NCO v8.7) with
	ESMTP id m1H6KEUi140130 for <cmm@us.ibm.com>; Sat, 16 Feb 2008 23:20:14
	-0700
Received: from d03av03.boulder.ibm.com (loopback [127.0.0.1]) by
	d03av03.boulder.ibm.com (8.12.11.20060308/8.13.3) with ESMTP id
	m1H6KEob025308 for <cmm@us.ibm.com>; Sat, 16 Feb 2008 23:20:14 -0700
Received: from d03as02.boulder.ibm.com (d03as02 [9.17.195.251]) by
	d03av03.boulder.ibm.com (8.12.11.20060308/8.12.11) with ESMTP id
	m1H6KElU025305 (version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-SHA bits=256
	verify=OK) for <cmm@us.ibm.com>; Sat, 16 Feb 2008 23:20:14 -0700
Received: from e32.co.us.ibm.com ([9.17.249.42]) by d03as02.boulder.ibm.com
	(8.12.11.20060308/8.12.11) with ESMTP id m1H6KCtt024992
	(version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-SHA bits=256 verify=OK) for
	<cmm@us.ibm.com>; Sat, 16 Feb 2008 23:20:13 -0700
Received: from wa-out-1112.google.com (wa-out-1112.google.com
	[209.85.146.178]) by e32.co.us.ibm.com (8.13.8/8.13.8) with ESMTP id
	m1H6JjCF030574 for <cmm@us.ibm.com>; Sun, 17 Feb 2008 01:20:04 -0500
Received: by wa-out-1112.google.com with SMTP id k34so2312165wah.10 for
	<cmm@us.ibm.com>; Sat, 16 Feb 2008 22:20:11 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;        d=gmail.com;
	s=gamma;
	h=domainkey-signature:received:received:date:from:to:cc:subject:message-id:mime-version:content-type:content-disposition:user-agent;
	       bh=pmQ/Rme4uwygvMy18gcJPWIDPn7FakLqbITEX5LqV/4=;       
	b=HOAlM+bUjGnaPJkLyOemMaIe2f+6fxeCriWWnqxf6b4fUXMU0EKpygkO/pisFRdMtXC+liXFTQmoRJNXDMg1SsWPDa3nXtAY5wdybhk/fcrqD+LOB6LcrP/I9VTK4UWaTqqIkALFBThipuaAXKJ9vEP3C7/sL9w1T6LARSg4RkY=
DomainKey-Signature: a=rsa-sha1; c=nofws;        d=gmail.com; s=gamma;
	h=date:from:to:cc:subject:message-id:mime-version:content-type:content-disposition:user-agent;
	      
	b=V8c3a/AkqeOtfCdigN3m82TzKcj4dstOw6pe2sV30aB60KIaGRWU2oq9DL/9n2YxWXQUO99wqCjWaox2618CSXQ7NrJ9IWwvWaWTTsIsT8Fq9m0D7xYnS3+rbCgZ4Wq8e6HZlnz4I4XR9kGr4LBvvCoR/FsqZEFIlNtZDI3jRHs=
Received: by 10.114.202.15 with SMTP id z15mr5108663waf.72.1203229211240;
	Sat, 16 Feb 2008 22:20:11 -0800 (PST)
Received: from @ ( [222.148.118.78]) by mx.google.com with ESMTPS id
	z20sm12797763pod.4.2008.02.16.22.20.08 (version=SSLv3 cipher=OTHER); Sat,
	16 Feb 2008 22:20:10 -0800 (PST)
Date: Sun, 17 Feb 2008 15:13:53 +0900
From: Akinobu Mita <akinobu.mita@gmail.com>
To: linux-ext4@vger.kernel.org
Cc: Andrew Morton <akpm@linux-foundation.org>,        Stephen Tweedie <sct@redhat.com>, adilger@clusterfs.com,        Mingming Cao <cmm@us.ibm.com>, Theodore Tso <tytso@MIT.EDU>
Subject: [PATCH 3/4] ext4: use ext4_get_group_desc()
Message-ID: <20080217061352.GJ3390@APFDCB5C>
MIME-Version: 1.0
Content-Type: text/plain; charset=iso-2022-jp
Content-Disposition: inline
User-Agent: Mutt/1.5.17 (2007-11-01)
X-Xagent-From: akinobu.mita@gmail.com
X-Xagent-To: cmm@linux.vnet.ibm.com
X-Xagent-Gateway: bldvmb.vnet.ibm.com (XAGENTU at BLDVMB)
X-Evolution-Source: imap://cmm@imap.linux.ibm.com/
Content-Transfer-Encoding: 8bit

Use ext4_get_group_desc()

Signed-off-by: Akinobu Mita <akinobu.mita@gmail.com>
Cc: Stephen Tweedie <sct@redhat.com>
Cc: adilger@clusterfs.com
Cc: Andrew Morton <akpm@linux-foundation.org>
Cc: Mingming Cao <cmm@us.ibm.com>
Cc: Theodore Tso <tytso@MIT.EDU>
---
 fs/ext4/inode.c |   20 +++-----------------
 1 file changed, 3 insertions(+), 17 deletions(-)

Index: linux-2.6.25-rc2/fs/ext4/inode.c
===================================================================
--- linux-2.6.25-rc2.orig/fs/ext4/inode.c	2008-02-18 23:25:16.000000000 -0800
+++ linux-2.6.25-rc2/fs/ext4/inode.c	2008-02-18 23:29:18.000000000 -0800
@@ -2498,12 +2498,10 @@ out_stop:
 static ext4_fsblk_t ext4_get_inode_block(struct super_block *sb,
 		unsigned long ino, struct ext4_iloc *iloc)
 {
-	unsigned long desc, group_desc;
 	ext4_group_t block_group;
 	unsigned long offset;
 	ext4_fsblk_t block;
-	struct buffer_head *bh;
-	struct ext4_group_desc * gdp;
+	struct ext4_group_desc *gdp;
 
 	if (!ext4_valid_inum(sb, ino)) {
 		/*
@@ -2515,22 +2513,10 @@ static ext4_fsblk_t ext4_get_inode_block
 	}
 
 	block_group = (ino - 1) / EXT4_INODES_PER_GROUP(sb);
-	if (block_group >= EXT4_SB(sb)->s_groups_count) {
-		ext4_error(sb,"ext4_get_inode_block","group >= groups count");
+	gdp = ext4_get_group_desc(sb, block_group, NULL);
+	if (!gdp)
 		return 0;
-	}
-	smp_rmb();
-	group_desc = block_group >> EXT4_DESC_PER_BLOCK_BITS(sb);
-	desc = block_group & (EXT4_DESC_PER_BLOCK(sb) - 1);
-	bh = EXT4_SB(sb)->s_group_desc[group_desc];
-	if (!bh) {
-		ext4_error (sb, "ext4_get_inode_block",
-			    "Descriptor not loaded");
-		return 0;
-	}
 
-	gdp = (struct ext4_group_desc *)((__u8 *)bh->b_data +
-		desc * EXT4_DESC_SIZE(sb));
 	/*
 	 * Figure out the offset within the block group inode table
 	 */
